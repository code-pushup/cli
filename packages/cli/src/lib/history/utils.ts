import { HistoryOptions } from '@code-pushup/core';
import { getHashFromTag, isSemver } from '@code-pushup/utils';
import { HistoryCliOptions } from './history.model';

export async function normalizeHashOptions(
  processArgs: HistoryCliOptions & HistoryOptions,
): Promise<HistoryCliOptions & HistoryOptions> {
  const {
    onlySemverTags,
    // overwritten
    maxCount,
    ...opt
  } = processArgs;

  // eslint-disable-next-line functional/no-let, prefer-const
  let { from, to, ...processOptions } = opt;
  // if no semver filter is used resolve hash of tags, as hashes are used to collect history
  if (!onlySemverTags) {
    if (from && isSemver(from)) {
      const { hash } = await getHashFromTag(from);
      from = hash;
    }
    if (to && isSemver(to)) {
      const { hash } = await getHashFromTag(to);
      to = hash;
    }
  }

  return {
    ...processOptions,
    onlySemverTags,
    maxCount: maxCount && maxCount > 0 ? maxCount : undefined,
    from,
    to,
  };
}
