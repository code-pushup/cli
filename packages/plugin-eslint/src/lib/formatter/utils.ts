import { ESLint } from 'eslint';
import { writeFileSync } from 'node:fs';
import { mkdir } from 'node:fs/promises';
import path from 'node:path';
import type { EslintFormat, FormatterConfig } from './types.js';

const eslint = new ESLint();

export function getExtensionForFormat(format: EslintFormat): string {
  const extensionMap: Record<EslintFormat, string> = {
    json: '.json',
    'json-with-metadata': 'json-with-metadata.json',
    html: 'html.html',
    xml: 'xml.xml',
    checkstyle: 'checkstyle.xml',
    junit: 'junit.xml',
    tap: '.tap',
    unix: 'unix.txt',
    stylish: 'stylish.txt',
    compact: 'compact.txt',
    codeframe: 'codeframe.txt',
    table: 'table.txt',
  };

  return extensionMap[format] || '.txt';
}

export async function formatContent(
  results: ESLint.LintResult[],
  format: EslintFormat,
): Promise<string> {
  const formatter = await eslint.loadFormatter(format);
  return formatter.format(results);
}

export function findConfigFromEnv(
  env: NodeJS.ProcessEnv,
): FormatterConfig | null {
  const configString = env['ESLINT_FORMATTER_CONFIG'];

  if (!configString || configString.trim() === '') {
    return null;
  }

  try {
    const parsed = JSON.parse(configString);

    // Validate that parsed result is an object
    if (typeof parsed !== 'object' || parsed == null || Array.isArray(parsed)) {
      console.error('ESLINT_FORMATTER_CONFIG must be a valid JSON object');
      return null;
    }

    return parsed as FormatterConfig;
  } catch (error_) {
    console.error(
      'Error parsing ESLINT_FORMATTER_CONFIG environment variable:',
      (error_ as Error).message,
    );
    return null;
  }
}

export async function handleTerminalOutput(
  format: EslintFormat | undefined,
  results: ESLint.LintResult[],
): Promise<void> {
  if (!format) {
    return;
  }
  const content = await formatContent(results, format);
  // eslint-disable-next-line no-console
  console.log(content);
}

export type PersistConfig = {
  outputDir: string;
  filename: string;
  format: EslintFormat;
  verbose?: boolean;
};

export async function persistEslintReport(
  content: string,
  options: PersistConfig,
): Promise<boolean> {
  const { outputDir, filename, format, verbose = false } = options;
  try {
    await mkdir(path.dirname(outputDir), { recursive: true });
    // eslint-disable-next-line n/no-sync
    writeFileSync(
      path.join(outputDir, `${filename}.${getExtensionForFormat(format)}`),
      content,
    );
    if (verbose) {
      // eslint-disable-next-line no-console
      console.log(`ESLint report (${format}) written to: ${outputDir}`);
    }
    return true;
  } catch (error_) {
    if (verbose) {
      console.error('There was a problem writing the output file:\n%s', error_);
    }
    return false;
  }
}

export async function persistEslintReports(
  formats: EslintFormat[],
  results: ESLint.LintResult[],
  options: { outputDir: string; filename: string; verbose: boolean },
): Promise<boolean> {
  const { outputDir, filename, verbose } = options;

  // eslint-disable-next-line functional/no-loop-statements
  for (const format of formats) {
    const content = await formatContent(results, format);

    const success = await persistEslintReport(content, {
      outputDir,
      filename,
      format,
      verbose,
    });

    if (!success) {
      return false;
    }
  }
  return true;
}
