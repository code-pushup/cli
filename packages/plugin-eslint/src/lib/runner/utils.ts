import type { ESLint } from 'eslint';
import { glob } from 'glob';
import type { PluginArtifactOptions } from '@code-pushup/models';
import {
  executeProcess,
  pluralizeToken,
  readJsonFile,
  ui,
} from '@code-pushup/utils';
import type { LinterOutput } from './types.js';

async function resolveGlobPatterns(patterns: string[]): Promise<string[]> {
  const resolvedPathArrays = await Promise.all(
    patterns.map(async pattern => glob(pattern)),
  );

  const resolvedPaths = resolvedPathArrays.flat();

  return [...new Set(resolvedPaths)].sort();
}

export async function loadArtifacts(
  artifacts: PluginArtifactOptions,
): Promise<LinterOutput[]> {
  if (artifacts.generateArtifactsCommand) {
    const { command, args = [] } =
      typeof artifacts.generateArtifactsCommand === 'string'
        ? { command: artifacts.generateArtifactsCommand }
        : artifacts.generateArtifactsCommand;

    // Log the actual command that was executed
    const commandString =
      typeof artifacts.generateArtifactsCommand === 'string'
        ? artifacts.generateArtifactsCommand
        : `${command} ${args.join(' ')}`;
    await ui().logger.log(`$ ${commandString}`);
    await executeProcess({
      command,
      args,
      ignoreExitCode: true,
    });
  }

  const initialArtifactPaths = Array.isArray(artifacts.artifactsPaths)
    ? artifacts.artifactsPaths
    : [artifacts.artifactsPaths];

  const artifactPaths = await resolveGlobPatterns(initialArtifactPaths);

  ui().logger.log(
    `ESLint plugin resolved ${initialArtifactPaths.length} ${pluralizeToken('pattern', initialArtifactPaths.length)} to ${artifactPaths.length} eslint ${pluralizeToken('report', artifactPaths.length)}`,
  );

  return await Promise.all(
    artifactPaths.map(async artifactPath => {
      // ESLint CLI outputs raw ESLint.LintResult[], but we need LinterOutput format
      const results = await readJsonFile<ESLint.LintResult[]>(artifactPath);
      return {
        results,
        ruleOptionsPerFile: {}, // TODO: This could be enhanced to load actual rule options if needed
      };
    }),
  );
}
