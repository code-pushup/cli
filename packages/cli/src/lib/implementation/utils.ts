import { toArray } from '@code-pushup/utils';

// log error and flush stdout so that Yargs doesn't suppress it
// related issue: https://github.com/yargs/yargs/issues/2118
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function logErrorBeforeThrow<T extends (...args: any[]) => any>(
  fn: T,
): T {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  return (async (...args: any[]) => {
    try {
      // eslint-disable-next-line @typescript-eslint/no-unsafe-return, @typescript-eslint/no-unsafe-argument
      return await fn(...args);
    } catch (error) {
      console.error(error);
      await new Promise(resolve => process.stdout.write('', resolve));
      throw error;
    }
  }) as T;
}

export function coerceArray<T extends string>(param: T | T[] = []): T[] {
  return [
    ...new Set(toArray(param).flatMap((f: T) => f.split(',') as T[]) || []),
  ];
}
