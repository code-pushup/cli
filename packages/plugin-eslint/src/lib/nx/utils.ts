import type { ProjectConfiguration } from '@nx/devkit';
import { join } from 'node:path';
import { fileExists, toArray } from '@code-pushup/utils';

export async function findCodePushupEslintrc(
  project: ProjectConfiguration,
): Promise<string | null> {
  const name = 'code-pushup.eslintrc';
  // https://eslint.org/docs/latest/use/configure/configuration-files#configuration-file-formats
  const extensions = ['json', 'js', 'cjs', 'yml', 'yaml'];

  // eslint-disable-next-line functional/no-loop-statements
  for (const ext of extensions) {
    const filename = `./${project.root}/${name}.${ext}`;
    if (await fileExists(join(process.cwd(), filename))) {
      return filename;
    }
  }

  return null;
}

export function getLintFilePatterns(project: ProjectConfiguration): string[] {
  const options = project.targets?.['lint']?.options as
    | { lintFilePatterns?: string | string[] }
    | undefined;
  return options?.lintFilePatterns == null
    ? [`${project.root}/**/*`] // lintFilePatterns defaults to ["{projectRoot}"] - https://github.com/nrwl/nx/pull/20313
    : toArray(options.lintFilePatterns);
}

export function getEslintConfig(
  project: ProjectConfiguration,
): string | undefined {
  const options = project.targets?.['lint']?.options as
    | { eslintConfig?: string }
    | undefined;
  return options?.eslintConfig ?? `./${project.root}/.eslintrc.json`;
}
