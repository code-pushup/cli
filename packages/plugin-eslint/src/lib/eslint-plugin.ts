import { mkdir, writeFile } from 'node:fs/promises';
import { dirname, join } from 'node:path';
import { fileURLToPath } from 'node:url';
import { PluginConfig } from '@code-pushup/models';
import { name, version } from '../../package.json';
import { ESLintPluginConfig, eslintPluginConfigSchema } from './config';
import { listAuditsAndGroups } from './meta';
import { ESLINTRC_PATH, createRunnerConfig } from './runner';
import { setupESLint } from './setup';

/**
 * Instantiates Code PushUp ESLint plugin for use in core config.
 *
 * @example
 * import eslintPlugin from '@code-pushup/eslint-plugin'
 *
 * export default {
 *   // ... core config ...
 *   plugins: [
 *     // ... other plugins ...
 *     await eslintPlugin({
 *       eslintrc: '.eslintrc.json',
 *       patterns: ['src', 'test/*.spec.js']
 *     })
 *   ]
 * }
 *
 * @param config Configuration options.
 * @returns Plugin configuration as a promise.
 */
export async function eslintPlugin(
  config: ESLintPluginConfig,
): Promise<PluginConfig> {
  const { eslintrc, patterns } = eslintPluginConfigSchema.parse(config);

  const eslint = setupESLint(eslintrc);

  const { audits, groups } = await listAuditsAndGroups(eslint, patterns);

  // save inline config to file so runner can access it later
  if (typeof eslintrc !== 'string') {
    await mkdir(dirname(ESLINTRC_PATH), { recursive: true });
    await writeFile(ESLINTRC_PATH, JSON.stringify(eslintrc));
  }
  const eslintrcPath = typeof eslintrc === 'string' ? eslintrc : ESLINTRC_PATH;

  const runnerScriptPath = join(
    fileURLToPath(dirname(import.meta.url)),
    'bin.js',
  );

  return {
    slug: 'eslint',
    title: 'ESLint',
    icon: 'eslint',
    description: 'Official Code PushUp ESLint plugin',
    docsUrl: 'https://www.npmjs.com/package/@code-pushup/eslint-plugin',
    packageName: name,
    version,

    audits,
    groups,

    runner: await createRunnerConfig(
      runnerScriptPath,
      audits,
      eslintrcPath,
      patterns,
    ),
  };
}
