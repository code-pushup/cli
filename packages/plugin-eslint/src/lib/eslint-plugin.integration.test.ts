import os from 'node:os';
import { dirname, join } from 'node:path';
import { fileURLToPath } from 'node:url';
import type { MockInstance } from 'vitest';
import type { Audit, PluginConfig, RunnerConfig } from '@code-pushup/models';
import { toUnixPath } from '@code-pushup/utils';
import { eslintPlugin } from './eslint-plugin';

describe('eslintPlugin', () => {
  const thisDir = fileURLToPath(dirname(import.meta.url));

  const fixturesDir = join(thisDir, '..', '..', 'mocks', 'fixtures');

  let cwdSpy: MockInstance<[], string>;
  let platformSpy: MockInstance<[], NodeJS.Platform>;

  const replaceAbsolutePath = (plugin: PluginConfig): PluginConfig => ({
    ...plugin,
    runner: {
      ...(plugin.runner as RunnerConfig),
      args: (plugin.runner as RunnerConfig).args?.map(arg =>
        toUnixPath(arg.replace(thisDir, '<dirname>')),
      ),
      outputFile: toUnixPath((plugin.runner as RunnerConfig).outputFile),
    },
  });

  beforeAll(() => {
    cwdSpy = vi.spyOn(process, 'cwd');
    // Linux produces extra quotation marks for globs
    platformSpy = vi.spyOn(os, 'platform').mockReturnValue('linux');
  });

  afterAll(() => {
    cwdSpy.mockRestore();
    platformSpy.mockRestore();
  });

  it('should initialize ESLint plugin for React application', async () => {
    cwdSpy.mockReturnValue(join(fixturesDir, 'todos-app'));
    const plugin = await eslintPlugin({
      eslintrc: '.eslintrc.js',
      patterns: ['src/**/*.js', 'src/**/*.jsx'],
    });

    expect(replaceAbsolutePath(plugin)).toMatchSnapshot({
      version: expect.any(String),
    });
  });

  it('should initialize ESLint plugin for Nx project', async () => {
    cwdSpy.mockReturnValue(join(fixturesDir, 'nx-monorepo'));
    const plugin = await eslintPlugin({
      eslintrc: './packages/utils/.eslintrc.json',
      patterns: ['packages/utils/**/*.ts', 'packages/utils/**/*.json'],
    });

    // expect rule from extended base .eslintrc.json
    expect(plugin.audits).toContainEqual(
      expect.objectContaining<Audit>({
        slug: expect.stringMatching(/^nx-enforce-module-boundaries/),
        title: expect.any(String),
        description: expect.stringContaining('sourceTag'),
      }),
    );
    // expect rule from utils project's .eslintrc.json
    expect(plugin.audits).toContainEqual(
      expect.objectContaining<Partial<Audit>>({
        slug: 'nx-dependency-checks',
      }),
    );
  });

  it('should initialize ESLint plugin using inline config', async () => {
    cwdSpy.mockReturnValue(thisDir);
    const plugin = await eslintPlugin({
      eslintrc: {
        extends: '@code-pushup',
      },
      patterns: '**/*.ts',
    });

    expect(plugin.groups?.length).toBeGreaterThanOrEqual(3);
    expect(plugin.audits.length).toBeGreaterThanOrEqual(200);
  });

  it('should throw when invalid parameters provided', async () => {
    await expect(
      // @ts-expect-error simulating invalid non-TS config
      eslintPlugin({ eslintrc: '.eslintrc.json' }),
    ).rejects.toThrow('patterns');
  });

  it("should throw if eslintrc file doesn't exist", async () => {
    await expect(
      eslintPlugin({ eslintrc: '.eslintrc.yml', patterns: '**/*.js' }),
    ).rejects.toThrow('Cannot read config file');
  });
});
