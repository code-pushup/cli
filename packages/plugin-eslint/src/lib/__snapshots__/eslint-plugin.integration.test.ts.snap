// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`eslintPlugin > should initialize ESLint plugin for React application 1`] = `
{
  "audits": [
    {
      "description": "ESLint rule **no-cond-assign**.",
      "docsUrl": "https://eslint.org/docs/latest/rules/no-cond-assign",
      "slug": "no-cond-assign",
      "title": "Disallow assignment operators in conditional expressions",
    },
    {
      "description": "ESLint rule **no-const-assign**.",
      "docsUrl": "https://eslint.org/docs/latest/rules/no-const-assign",
      "slug": "no-const-assign",
      "title": "Disallow reassigning \`const\` variables",
    },
    {
      "description": "ESLint rule **no-debugger**.",
      "docsUrl": "https://eslint.org/docs/latest/rules/no-debugger",
      "slug": "no-debugger",
      "title": "Disallow the use of \`debugger\`",
    },
    {
      "description": "ESLint rule **no-invalid-regexp**.",
      "docsUrl": "https://eslint.org/docs/latest/rules/no-invalid-regexp",
      "slug": "no-invalid-regexp",
      "title": "Disallow invalid regular expression strings in \`RegExp\` constructors",
    },
    {
      "description": "ESLint rule **no-undef**.",
      "docsUrl": "https://eslint.org/docs/latest/rules/no-undef",
      "slug": "no-undef",
      "title": "Disallow the use of undeclared variables unless mentioned in \`/*global */\` comments",
    },
    {
      "description": "ESLint rule **no-unreachable-loop**.",
      "docsUrl": "https://eslint.org/docs/latest/rules/no-unreachable-loop",
      "slug": "no-unreachable-loop",
      "title": "Disallow loops with a body that allows only one iteration",
    },
    {
      "description": "ESLint rule **no-unsafe-negation**.",
      "docsUrl": "https://eslint.org/docs/latest/rules/no-unsafe-negation",
      "slug": "no-unsafe-negation",
      "title": "Disallow negating the left operand of relational operators",
    },
    {
      "description": "ESLint rule **no-unsafe-optional-chaining**.",
      "docsUrl": "https://eslint.org/docs/latest/rules/no-unsafe-optional-chaining",
      "slug": "no-unsafe-optional-chaining",
      "title": "Disallow use of optional chaining in contexts where the \`undefined\` value is not allowed",
    },
    {
      "description": "ESLint rule **no-unused-vars**.",
      "docsUrl": "https://eslint.org/docs/latest/rules/no-unused-vars",
      "slug": "no-unused-vars",
      "title": "Disallow unused variables",
    },
    {
      "description": "ESLint rule **use-isnan**.",
      "docsUrl": "https://eslint.org/docs/latest/rules/use-isnan",
      "slug": "use-isnan",
      "title": "Require calls to \`isNaN()\` when checking for \`NaN\`",
    },
    {
      "description": "ESLint rule **valid-typeof**.",
      "docsUrl": "https://eslint.org/docs/latest/rules/valid-typeof",
      "slug": "valid-typeof",
      "title": "Enforce comparing \`typeof\` expressions against valid strings",
    },
    {
      "description": "ESLint rule **arrow-body-style**.",
      "docsUrl": "https://eslint.org/docs/latest/rules/arrow-body-style",
      "slug": "arrow-body-style",
      "title": "Require braces around arrow function bodies",
    },
    {
      "description": "ESLint rule **camelcase**.",
      "docsUrl": "https://eslint.org/docs/latest/rules/camelcase",
      "slug": "camelcase",
      "title": "Enforce camelcase naming convention",
    },
    {
      "description": "ESLint rule **curly**.",
      "docsUrl": "https://eslint.org/docs/latest/rules/curly",
      "slug": "curly",
      "title": "Enforce consistent brace style for all control statements",
    },
    {
      "description": "ESLint rule **eqeqeq**.",
      "docsUrl": "https://eslint.org/docs/latest/rules/eqeqeq",
      "slug": "eqeqeq",
      "title": "Require the use of \`===\` and \`!==\`",
    },
    {
      "description": "ESLint rule **max-lines-per-function**.",
      "docsUrl": "https://eslint.org/docs/latest/rules/max-lines-per-function",
      "slug": "max-lines-per-function",
      "title": "Enforce a maximum number of lines of code in a function",
    },
    {
      "description": "ESLint rule **max-lines**.",
      "docsUrl": "https://eslint.org/docs/latest/rules/max-lines",
      "slug": "max-lines",
      "title": "Enforce a maximum number of lines per file",
    },
    {
      "description": "ESLint rule **no-shadow**.",
      "docsUrl": "https://eslint.org/docs/latest/rules/no-shadow",
      "slug": "no-shadow",
      "title": "Disallow variable declarations from shadowing variables declared in the outer scope",
    },
    {
      "description": "ESLint rule **no-var**.",
      "docsUrl": "https://eslint.org/docs/latest/rules/no-var",
      "slug": "no-var",
      "title": "Require \`let\` or \`const\` instead of \`var\`",
    },
    {
      "description": "ESLint rule **object-shorthand**.",
      "docsUrl": "https://eslint.org/docs/latest/rules/object-shorthand",
      "slug": "object-shorthand",
      "title": "Require or disallow method and property shorthand syntax for object literals",
    },
    {
      "description": "ESLint rule **prefer-arrow-callback**.",
      "docsUrl": "https://eslint.org/docs/latest/rules/prefer-arrow-callback",
      "slug": "prefer-arrow-callback",
      "title": "Require using arrow functions for callbacks",
    },
    {
      "description": "ESLint rule **prefer-const**.",
      "docsUrl": "https://eslint.org/docs/latest/rules/prefer-const",
      "slug": "prefer-const",
      "title": "Require \`const\` declarations for variables that are never reassigned after declared",
    },
    {
      "description": "ESLint rule **prefer-object-spread**.",
      "docsUrl": "https://eslint.org/docs/latest/rules/prefer-object-spread",
      "slug": "prefer-object-spread",
      "title": "Disallow using Object.assign with an object literal as the first argument and prefer the use of object spread instead",
    },
    {
      "description": "ESLint rule **yoda**.",
      "docsUrl": "https://eslint.org/docs/latest/rules/yoda",
      "slug": "yoda",
      "title": "Require or disallow "Yoda" conditions",
    },
    {
      "description": "ESLint rule **jsx-key**, from _react_ plugin.",
      "docsUrl": "https://github.com/jsx-eslint/eslint-plugin-react/tree/master/docs/rules/jsx-key.md",
      "slug": "react-jsx-key",
      "title": "Disallow missing \`key\` props in iterators/collection literals",
    },
    {
      "description": "ESLint rule **prop-types**, from _react_ plugin.",
      "docsUrl": "https://github.com/jsx-eslint/eslint-plugin-react/tree/master/docs/rules/prop-types.md",
      "slug": "react-prop-types",
      "title": "Disallow missing props validation in a React component definition",
    },
    {
      "description": "ESLint rule **react-in-jsx-scope**, from _react_ plugin.",
      "docsUrl": "https://github.com/jsx-eslint/eslint-plugin-react/tree/master/docs/rules/react-in-jsx-scope.md",
      "slug": "react-react-in-jsx-scope",
      "title": "Disallow missing React when using JSX",
    },
    {
      "description": "ESLint rule **rules-of-hooks**, from _react-hooks_ plugin.",
      "docsUrl": "https://reactjs.org/docs/hooks-rules.html",
      "slug": "react-hooks-rules-of-hooks",
      "title": "enforces the Rules of Hooks",
    },
    {
      "description": "ESLint rule **exhaustive-deps**, from _react-hooks_ plugin.",
      "docsUrl": "https://github.com/facebook/react/issues/14920",
      "slug": "react-hooks-exhaustive-deps",
      "title": "verifies the list of dependencies for Hooks like useEffect and similar",
    },
    {
      "description": "ESLint rule **display-name**, from _react_ plugin.",
      "docsUrl": "https://github.com/jsx-eslint/eslint-plugin-react/tree/master/docs/rules/display-name.md",
      "slug": "react-display-name",
      "title": "Disallow missing displayName in a React component definition",
    },
    {
      "description": "ESLint rule **jsx-no-comment-textnodes**, from _react_ plugin.",
      "docsUrl": "https://github.com/jsx-eslint/eslint-plugin-react/tree/master/docs/rules/jsx-no-comment-textnodes.md",
      "slug": "react-jsx-no-comment-textnodes",
      "title": "Disallow comments from being inserted as text nodes",
    },
    {
      "description": "ESLint rule **jsx-no-duplicate-props**, from _react_ plugin.",
      "docsUrl": "https://github.com/jsx-eslint/eslint-plugin-react/tree/master/docs/rules/jsx-no-duplicate-props.md",
      "slug": "react-jsx-no-duplicate-props",
      "title": "Disallow duplicate properties in JSX",
    },
    {
      "description": "ESLint rule **jsx-no-target-blank**, from _react_ plugin.",
      "docsUrl": "https://github.com/jsx-eslint/eslint-plugin-react/tree/master/docs/rules/jsx-no-target-blank.md",
      "slug": "react-jsx-no-target-blank",
      "title": "Disallow \`target="_blank"\` attribute without \`rel="noreferrer"\`",
    },
    {
      "description": "ESLint rule **jsx-no-undef**, from _react_ plugin.",
      "docsUrl": "https://github.com/jsx-eslint/eslint-plugin-react/tree/master/docs/rules/jsx-no-undef.md",
      "slug": "react-jsx-no-undef",
      "title": "Disallow undeclared variables in JSX",
    },
    {
      "description": "ESLint rule **jsx-uses-react**, from _react_ plugin.",
      "docsUrl": "https://github.com/jsx-eslint/eslint-plugin-react/tree/master/docs/rules/jsx-uses-react.md",
      "slug": "react-jsx-uses-react",
      "title": "Disallow React to be incorrectly marked as unused",
    },
    {
      "description": "ESLint rule **jsx-uses-vars**, from _react_ plugin.",
      "docsUrl": "https://github.com/jsx-eslint/eslint-plugin-react/tree/master/docs/rules/jsx-uses-vars.md",
      "slug": "react-jsx-uses-vars",
      "title": "Disallow variables used in JSX to be incorrectly marked as unused",
    },
    {
      "description": "ESLint rule **no-children-prop**, from _react_ plugin.",
      "docsUrl": "https://github.com/jsx-eslint/eslint-plugin-react/tree/master/docs/rules/no-children-prop.md",
      "slug": "react-no-children-prop",
      "title": "Disallow passing of children as props",
    },
    {
      "description": "ESLint rule **no-danger-with-children**, from _react_ plugin.",
      "docsUrl": "https://github.com/jsx-eslint/eslint-plugin-react/tree/master/docs/rules/no-danger-with-children.md",
      "slug": "react-no-danger-with-children",
      "title": "Disallow when a DOM element is using both children and dangerouslySetInnerHTML",
    },
    {
      "description": "ESLint rule **no-deprecated**, from _react_ plugin.",
      "docsUrl": "https://github.com/jsx-eslint/eslint-plugin-react/tree/master/docs/rules/no-deprecated.md",
      "slug": "react-no-deprecated",
      "title": "Disallow usage of deprecated methods",
    },
    {
      "description": "ESLint rule **no-direct-mutation-state**, from _react_ plugin.",
      "docsUrl": "https://github.com/jsx-eslint/eslint-plugin-react/tree/master/docs/rules/no-direct-mutation-state.md",
      "slug": "react-no-direct-mutation-state",
      "title": "Disallow direct mutation of this.state",
    },
    {
      "description": "ESLint rule **no-find-dom-node**, from _react_ plugin.",
      "docsUrl": "https://github.com/jsx-eslint/eslint-plugin-react/tree/master/docs/rules/no-find-dom-node.md",
      "slug": "react-no-find-dom-node",
      "title": "Disallow usage of findDOMNode",
    },
    {
      "description": "ESLint rule **no-is-mounted**, from _react_ plugin.",
      "docsUrl": "https://github.com/jsx-eslint/eslint-plugin-react/tree/master/docs/rules/no-is-mounted.md",
      "slug": "react-no-is-mounted",
      "title": "Disallow usage of isMounted",
    },
    {
      "description": "ESLint rule **no-render-return-value**, from _react_ plugin.",
      "docsUrl": "https://github.com/jsx-eslint/eslint-plugin-react/tree/master/docs/rules/no-render-return-value.md",
      "slug": "react-no-render-return-value",
      "title": "Disallow usage of the return value of ReactDOM.render",
    },
    {
      "description": "ESLint rule **no-string-refs**, from _react_ plugin.",
      "docsUrl": "https://github.com/jsx-eslint/eslint-plugin-react/tree/master/docs/rules/no-string-refs.md",
      "slug": "react-no-string-refs",
      "title": "Disallow using string references",
    },
    {
      "description": "ESLint rule **no-unescaped-entities**, from _react_ plugin.",
      "docsUrl": "https://github.com/jsx-eslint/eslint-plugin-react/tree/master/docs/rules/no-unescaped-entities.md",
      "slug": "react-no-unescaped-entities",
      "title": "Disallow unescaped HTML entities from appearing in markup",
    },
    {
      "description": "ESLint rule **no-unknown-property**, from _react_ plugin.",
      "docsUrl": "https://github.com/jsx-eslint/eslint-plugin-react/tree/master/docs/rules/no-unknown-property.md",
      "slug": "react-no-unknown-property",
      "title": "Disallow usage of unknown DOM property",
    },
    {
      "description": "ESLint rule **require-render-return**, from _react_ plugin.",
      "docsUrl": "https://github.com/jsx-eslint/eslint-plugin-react/tree/master/docs/rules/require-render-return.md",
      "slug": "react-require-render-return",
      "title": "Enforce ES5 or ES6 class for returning value in render function",
    },
  ],
  "description": "Official Code PushUp ESLint plugin",
  "docsUrl": "https://www.npmjs.com/package/@code-pushup/eslint-plugin",
  "groups": [
    {
      "description": "Code that either will cause an error or may cause confusing behavior. Developers should consider this a high priority to resolve.",
      "refs": [
        {
          "slug": "no-cond-assign",
          "weight": 1,
        },
        {
          "slug": "no-const-assign",
          "weight": 1,
        },
        {
          "slug": "no-debugger",
          "weight": 1,
        },
        {
          "slug": "no-invalid-regexp",
          "weight": 1,
        },
        {
          "slug": "no-undef",
          "weight": 1,
        },
        {
          "slug": "no-unreachable-loop",
          "weight": 1,
        },
        {
          "slug": "no-unsafe-negation",
          "weight": 1,
        },
        {
          "slug": "no-unsafe-optional-chaining",
          "weight": 1,
        },
        {
          "slug": "no-unused-vars",
          "weight": 1,
        },
        {
          "slug": "use-isnan",
          "weight": 1,
        },
        {
          "slug": "valid-typeof",
          "weight": 1,
        },
        {
          "slug": "react-hooks-rules-of-hooks",
          "weight": 1,
        },
      ],
      "slug": "problems",
      "title": "Problems",
    },
    {
      "description": "Something that could be done in a better way but no errors will occur if the code isn't changed.",
      "refs": [
        {
          "slug": "arrow-body-style",
          "weight": 1,
        },
        {
          "slug": "camelcase",
          "weight": 1,
        },
        {
          "slug": "curly",
          "weight": 1,
        },
        {
          "slug": "eqeqeq",
          "weight": 1,
        },
        {
          "slug": "max-lines-per-function",
          "weight": 1,
        },
        {
          "slug": "max-lines",
          "weight": 1,
        },
        {
          "slug": "no-shadow",
          "weight": 1,
        },
        {
          "slug": "no-var",
          "weight": 1,
        },
        {
          "slug": "object-shorthand",
          "weight": 1,
        },
        {
          "slug": "prefer-arrow-callback",
          "weight": 1,
        },
        {
          "slug": "prefer-const",
          "weight": 1,
        },
        {
          "slug": "prefer-object-spread",
          "weight": 1,
        },
        {
          "slug": "yoda",
          "weight": 1,
        },
        {
          "slug": "react-hooks-exhaustive-deps",
          "weight": 1,
        },
      ],
      "slug": "suggestions",
      "title": "Suggestions",
    },
    {
      "refs": [
        {
          "slug": "react-prop-types",
          "weight": 1,
        },
        {
          "slug": "react-display-name",
          "weight": 1,
        },
        {
          "slug": "react-jsx-no-target-blank",
          "weight": 1,
        },
        {
          "slug": "react-jsx-uses-react",
          "weight": 1,
        },
        {
          "slug": "react-jsx-uses-vars",
          "weight": 1,
        },
        {
          "slug": "react-no-children-prop",
          "weight": 1,
        },
        {
          "slug": "react-no-deprecated",
          "weight": 1,
        },
        {
          "slug": "react-no-find-dom-node",
          "weight": 1,
        },
        {
          "slug": "react-no-is-mounted",
          "weight": 1,
        },
        {
          "slug": "react-no-render-return-value",
          "weight": 1,
        },
        {
          "slug": "react-no-string-refs",
          "weight": 1,
        },
      ],
      "slug": "react-best-practices",
      "title": "Best Practices (react)",
    },
    {
      "refs": [
        {
          "slug": "react-jsx-key",
          "weight": 1,
        },
        {
          "slug": "react-react-in-jsx-scope",
          "weight": 1,
        },
        {
          "slug": "react-jsx-no-comment-textnodes",
          "weight": 1,
        },
        {
          "slug": "react-jsx-no-duplicate-props",
          "weight": 1,
        },
        {
          "slug": "react-jsx-no-undef",
          "weight": 1,
        },
        {
          "slug": "react-no-danger-with-children",
          "weight": 1,
        },
        {
          "slug": "react-no-direct-mutation-state",
          "weight": 1,
        },
        {
          "slug": "react-no-unescaped-entities",
          "weight": 1,
        },
        {
          "slug": "react-no-unknown-property",
          "weight": 1,
        },
        {
          "slug": "react-require-render-return",
          "weight": 1,
        },
      ],
      "slug": "react-possible-errors",
      "title": "Possible Errors (react)",
    },
  ],
  "icon": "eslint",
  "packageName": "@code-pushup/eslint-plugin",
  "runner": {
    "args": [
      ""<dirname>/bin.js"",
    ],
    "command": "node",
    "outputFile": "node_modules/.code-pushup/eslint/runner-output.json",
  },
  "slug": "eslint",
  "title": "ESLint",
  "version": Any<String>,
}
`;
